algorithms/vi-map-helpers/src/vi-map-landmark-quality-evaluation.cc

```diff
@@ -91,4 +144,175 @@ void resetLandmarkQualityToUnknown(
+      observations_with_track_id;
+  for (const auto& keypoint : landmark_observations) {
+    vi_map::KeypointIdentifierList& frame_observation =
+        frames_with_observations[keypoint.frame_id];
+    frame_observation.push_back(keypoint);
+    const pose_graph::VertexId& vertex_id = keypoint.frame_id.vertex_id;
+    CHECK(vertex_id.isValid());
+    CHECK(map.hasVertex(vertex_id));
+    const vi_map::Vertex& vertex = map.getVertex(vertex_id);
+    CHECK(vertex.hasVisualNFrame());
+    CHECK(vertex.getVisualFrame(keypoint.frame_id.frame_index).isValid());
+    const int track_id = vertex.getVisualFrame(keypoint.frame_id.frame_index)
+                             .getTrackId(keypoint.keypoint_index);
+    vi_map::KeypointIdentifierList& track_id_observations =
+        track_ids_with_observations[track_id];
+    track_id_observations.push_back(keypoint);
+    observations_with_track_id[keypoint] = track_id;
+  }
+
+  std::vector<int> outlier_track_ids;
+  static constexpr int kMaxNumSameLandmarkId = 1;
+  for (auto it = frames_with_observations.begin();
+       it != frames_with_observations.end(); ++it) {
+    // If the landmark is only observed once nothing has to be done.
+    if (it->second.size() <= kMaxNumSameLandmarkId) {
+      continue;
+    }
+
+    const vi_map::Vertex& vertex = map.getVertex(it->first.vertex_id);
+    const Eigen::Vector3d& p_C_fi =
+        map.getLandmark_p_C_fi(landmark.id(), vertex, it->first.frame_index);
+
+    vi_map::KeypointIdentifier min_error_keypoint;
+
+    // If landmark lies behind the camera plane we remove all observations
+    // from this frame from the landmark. Otherwise we keep the observation
+    // with the smallest reprojection error.
+    if (p_C_fi[2] > 0.0) {
+      double min_sq_reprojection_error = std::numeric_limits<double>::max();
+      for (const vi_map::KeypointIdentifier& keypoint : it->second) {
+        const double reprojection_error_sq = computeSquaredReprojectionError(
+            vertex, keypoint.frame_id.frame_index, keypoint.keypoint_index,
+            p_C_fi);
+        if (reprojection_error_sq < min_sq_reprojection_error) {
+          min_error_keypoint = keypoint;
+          min_sq_reprojection_error = reprojection_error_sq;
+        }
+      }
+    }
+
+    for (const vi_map::KeypointIdentifier& keypoint : it->second) {
+      if (keypoint == min_error_keypoint) {
+        continue;
+      }
+      // Mark the track ids of the observations with non minimal
+      // reprojection error as outliers
+      if (std::find(
+              outlier_track_ids.begin(), outlier_track_ids.end(),
+              observations_with_track_id[keypoint]) ==
+          outlier_track_ids.end()) {
+        outlier_track_ids.push_back(observations_with_track_id[keypoint]);
+      }
+    }
+  }
+
+  // Store the outlier track ids with the respective observations
+  for (const int outlier_track_id : outlier_track_ids) {
+    (*tracks_with_keypoints)[outlier_track_id] =
```

Comment:
Might this not present a problem? Like I can see maybe one track having multiple wrong keypoint associations maybe? And then here you would only store one instead of all of them. If this is already done entirely in a non parallel way why not move immediately to detaching something when we find it a bit higher up in the code? Instead of trying to store a complicated list of stuff without clashes in a parallel fashion